generate kernels for codegend out= operators

bdhirsh opened the pull request:
We already generate out= variants of NativeFunctions for all (non-CompositeImplicitAutograd) ops today, but the main benefit was to be able to properly group them up in the codegen, so we didn't have kernels for them. This PR generates out= kernels for all of the existing code-generated out= NativeFunctions.

Martin mentioned that having working kernels for all out= ops will be useful to mobile for development (you can run a mobile model E2E, without having 100% out= custom kernel coverage)

The kernels are generated by calling into their functional versions, and coping the results back into the out= arguments.

**Testing?**
Testing is kind of hard, because these kernels aren't actually used anywhere in core, and aren't exposed to python. I just stared at the output and made sure it looked reasonable (we properly resize and copy the output), but if anyone is concerned / has suggestions on tests I could add some.

some example outputs:
```
at::Tensor & add_out(const at::Tensor & self, const at::Scalar & other, const at::Scalar & alpha, at::Tensor & out) {
  auto tmp_output = at::_ops::add_Scalar::call(self, other, alpha);
    resize_out_helper(out, tmp_output);
  copy_arg(out, tmp_output);
  return out;
}


void _foreach_add_out(at::TensorList self, const at::Scalar & scalar, at::TensorList out) {
  auto tmp_output = at::_ops::_foreach_add_Scalar_functional::call(self, scalar);
    resize_out_helper(out, tmp_output);
  copy_arg(out, tmp_output);

}


::std::tuple<at::Tensor &,at::Tensor &> _fused_moving_avg_obs_fq_helper_out(const at::Tensor & self, const at::Tensor & observer_on, const at::Tensor & fake_quant_on, at::Tens
  auto tmp_output = at::_ops::_fused_moving_avg_obs_fq_helper_functional::call(self, observer_on, fake_quant_on, running_min, running_max, scale, zero_point, averaging_const,
    resize_out_helper(out0, std::get<0>(tmp_output));
  copy_arg(out0, std::get<0>(tmp_output));
  resize_out_helper(out1, std::get<1>(tmp_output));
  copy_arg(out1, std::get<1>(tmp_output));
  return ::std::tuple<at::Tensor &,at::Tensor &>(out0, out1);
}
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #78626

bdhirsh committed:
generate kernels for codegend out= operators

[ghstack-poisoned]

JacobSzwejbka commented:
void _foreach_add_out(at::TensorList self, const at::Scalar & scalar, at::TensorList out) {
  auto tmp_output = at::_ops::_foreach_add_Scalar_functional::call(self, scalar);
  copy_arg(out, tmp_output);
}
What happens if out and tmp_output are not the same size list? throw?

::std::tuple<at::Tensor &,at::Tensor &> _fused_moving_avg_obs_fq_helper_out(const at::Tensor & self, const at::Tensor & observer_on, const at::Tensor & fake_quant_on, at::Tens
  auto tmp_output = at::_ops::_fused_moving_avg_obs_fq_helper_functional::call(self, observer_on, fake_quant_on, running_min, running_max, scale, zero_point, averaging_const,
  copy_arg(out0, std::get<0>(tmp_output));
copy_arg(out1, std::get<1>(tmp_output));
  return ::std::tuple<at::Tensor &,at::Tensor &>(out0, out1);
}
Hmm this one is interesting. From the return it looks like the elements will be seperate in the input schema, for executorch serialization I guess will have to create a list for us whose elements are out0 and out1 somehow

bdhirsh commented:
> What happens if out and tmp_output are not the same size list? throw?

Right now it'll throw, yep. At least today, I don't think there are any tensorlist out= ops that are exposed to python. These will really just be used in memory planning passes. So we don't need to bend over backwards to make sure that the "right thing" always happens - we should just expect the code doing the memory planning to know what sized out= tensorlist to create.

> Hmm this one is interesting. From the return it looks like the elements will be seperate in the input schema, for executorch serialization I guess will have to create a list for us whose elements are out0 and out1 somehow

Oh, is the idea that executorch will need to potentially special-case `Tuple<Tensor, Tensor>` outputs, in the same way you need special logic for `TensorList` today?

JacobSzwejbka commented:
> > Oh, is the idea that executorch will need to potentially special-case `Tuple<Tensor, Tensor>` outputs, in the same way you need special logic for `TensorList` today?
> 
> Yeah my thought is if an operator returns a tuple, and that tuple is then used in another op it would have to be passed in as tensorlist right? so we will need to know to have an EValue in the plan ahead of time that is that list, and have its elements be the out elements passed in separately to this function. It might be weird on the compiler side to deal with this, but the runtime shouldnt have to worry about it I dont think. We wont consume the return value of an out variant op.

bdhirsh committed:
Update on "generate kernels for codegend out= operators"

We already generate out= variants of NativeFunctions for all (non-CompositeImplicitAutograd) ops today, but the main benefit was to be able to properly group them up in the codegen, so we didn't have kernels for them. This PR generates out= kernels for all of the existing code-generated out= NativeFunctions.

Martin mentioned that having working kernels for all out= ops will be useful to mobile for development, even if the generated kernels aren't 100% efficient.

The kernels are generated by calling into their functional versions, and coping the results back into the out= arguments.


some example outputs:
```
at::Tensor & add_out(const at::Tensor & self, const at::Scalar & other, const at::Scalar & alpha, at::Tensor & out) {
  auto tmp_output = at::_ops::add_Scalar::call(self, other, alpha);
    resize_out_helper(out, tmp_output);
  copy_arg(out, tmp_output);
  return out;
}


void _foreach_add_out(at::TensorList self, const at::Scalar & scalar, at::TensorList out) {
  auto tmp_output = at::_ops::_foreach_add_Scalar_functional::call(self, scalar);
    resize_out_helper(out, tmp_output);
  copy_arg(out, tmp_output);

}


::std::tuple<at::Tensor &,at::Tensor &> _fused_moving_avg_obs_fq_helper_out(const at::Tensor & self, const at::Tensor & observer_on, const at::Tensor & fake_quant_on, at::Tens
  auto tmp_output = at::_ops::_fused_moving_avg_obs_fq_helper_functional::call(self, observer_on, fake_quant_on, running_min, running_max, scale, zero_point, averaging_const,
    resize_out_helper(out0, std::get<0>(tmp_output));
  copy_arg(out0, std::get<0>(tmp_output));
  resize_out_helper(out1, std::get<1>(tmp_output));
  copy_arg(out1, std::get<1>(tmp_output));
  return ::std::tuple<at::Tensor &,at::Tensor &>(out0, out1);
}
```




[ghstack-poisoned]

bdhirsh commented:
> Yeah my thought is if an operator returns a tuple, and that tuple is then used in another op

That's technically possible if someone is writing arbitrary C++ code. But with executorch, you only need to worry about the final code that is spit out of tracing a python model and running memory planning, right? If that's true, then it might be easier for the out= conversion step to ensure that it always extracts individual tensors out of tuples before passing them as arguments, so you just don't have to worry about that case

bdhirsh commented:
(this is ready for review btw)

JacobSzwejbka commented:
Looks good to me cc @zhxchen17 to see if the tuple and list out variants look workable from the compiler perspective

larryliu0820 commented on aten/src/ATen/templates/CompositeViewCopyKernels.cpp:
```diff
@@ -28,10 +29,34 @@ std::vector<at::Tensor> clone_arg(const at::TensorList& t_list) {
+void copy_arg(const at::TensorList& dst, const at::TensorList& src) {
+    TORCH_INTERNAL_ASSERT(dst.size() == src.size());
+    for (const auto& i : c10::irange(dst.size())) {
+        dst[i].copy_(src[i]);
```
nit: is it better to call `copy_arg()` here?

larryliu0820 commented on aten/src/ATen/templates/CompositeViewCopyKernels.cpp:
```diff
@@ -28,10 +29,34 @@ std::vector<at::Tensor> clone_arg(const at::TensorList& t_list) {
+void resize_out_helper(const at::TensorList& dst, const at::TensorList& src) {
+    TORCH_INTERNAL_ASSERT(dst.size() == src.size());
+    for (const auto& i : c10::irange(dst.size())) {
+        at::native::resize_output(dst[i], src[i].sizes());
```
Same here

zhxchen17 commented:
> Looks good to me cc @zhxchen17 to see if the tuple and list out variants look workable from the compiler perspective

@JacobSzwejbka Yeah I agree with Brian. Basically we still need TensorList as input argument because we don't know the number of arguments during codegen. For tuple of tensor, we can just flatten & inline them in the arg list, and I think we already do that in compiler (if not I can add it).
